### DISCUSSION TITLE
Support TypeScript Project

### DISCUSSION THREAD

**Discussion by jellydn** [deta/cloud-docs#226] [Ideas] [answered]
Hi there,

I've checked the source code and that Deta Micro only supports `index.js`.

https://github.com/deta/deta-cli/blob/master/runtime/manager.go#L50-L54

Could we support `index.ts` or something else like `server.ts` or even more customized in the future? Thank you very much.

**Comment by abdelhai** [comment, marked as answer]
hi, this is a limitation of our current runtime, not the cli. we're working on a new runtime with the aim to support TS.

**Reply by HanzCEO** [reply]
In the meantime, users could precompile their ts project before pushing to Deta.

Or, use npm scripts like `postinstall` or `build`.

**Reply by ngdangtu-vn** [reply]
In my project, my repo structure separates ts code (as `/src`) and js code (as `/server`). And the thing is I usually remove all the /server dir. The Deta CLI always put their own dir in the same level of entrypoint which I think that I have to do `deta new` all the time which I do not like it.

**Reply by abdelhai** [reply]
you can copy the `.deta` dir or keep it somewhere. `deta new` creates a new micro, which is not what you should want. you could always do `deta clone...` to get a specific `.deta` dir.

sorry for the messy workaround, but we're working on supporting more complex project structures atm.

**Comment by ArnabXD** [comment]
> In the meantime, users could precompile their ts project before pushing to Deta.
> 
> Or, use npm scripts like `postinstall` or `build`.

@HanzCEO  can you share a valid tsconfig ? Because I have tried this long ago and the transpiled js code doesn't work as expected. throwing errors for things like optional chaining.

**Reply by bogdaaamn** [reply]
This is the one I use with `tsc`: [deta-typescript-express-starter/tsconfig.json](https://github.com/BogDAAAMN/deta-typescript-express-starter/blob/9a0685d673c67c985f46c6db85dd3d0164e10659/tsconfig.json#L1-L11)

**Comment by AndreWohnsland** [comment]
Any updates on this topic? I also stumbled on this issue, because I am currently in the process of migrating one of my dyno from heroku to deta. Usually after pushing a change, my postinstall just compiles the ts and everything is defined in the package.json, but for deta there is some workaround necessary to make it work (similar as describe by others). Would love to be able to use .ts files and jsut the defined pipeline for compiling on the run.

**Comment by bogdaaamn** [comment]
Same problem here but I managed to make it work with `tsc`. 

I made a little [deta-typescript-express-starter](https://github.com/BogDAAAMN/deta-typescript-express-starter) template, following it, I write all the Typescript in `src/` and compile it all in `dist/`. Then I import `dist/index.js` into `index.js` and export it and it's all picked up by Deta, check [`index.js`](https://github.com/BogDAAAMN/deta-typescript-express-starter/blob/main/index.js). This is the most simple [`tsconfig.json`](https://github.com/BogDAAAMN/deta-typescript-express-starter/blob/main/tsconfig.json).

I guess the easiest solution is to compile all the `.ts` files into `index.js`, but I didn't manage to make `requireJS` work for some reason so I did it a bit messier, but I will find the solution one day 😂
